"NAME\n"
"====\n"
"\n"
"intronProspector â€” Extract putative junctions from RNA-Seq alignments\n"
"\n"
"SYNOPSIS\n"
"========\n"
"\n"
"intronProspector [options] [readaligns]\n"
"\n"
"DESCRIPTION\n"
"===========\n"
"\n"
"Find putative intron junctions in a RNA-Seq alignment. The readaligns\n"
"file maybe in SAM, BAM, or CRAM format and does not need to be sorted or\n"
"indexed and maybe streamed.\n"
"\n"
"This program allows for calling splice junctions indebtedness of the\n"
"alignment program. It maybe used in a pipeline, copying the alignment\n"
"file on stdin to stdout. It can sit between an aligner outputting SAM\n"
"and samtools to convert to BAM/CRAM.\n"
"\n"
"Options\n"
"-------\n"
"\n"
"-h, --help\n"
"\n"
"  Prints brief usage information.\n"
"\n"
"-v, --version\n"
"\n"
"  Prints the current version number.\n"
"\n"
"-a INT, --min-anchor-length=INT\n"
"\n"
"  Minimum anchor length. Junctions which satisfy a minimum anchor length\n"
"  on both sides are reported. Mismatch bases don't count towards the\n"
"  meeting this threshold. The default is 8 bases.\n"
"\n"
"-i INT, --min-intron_length=INT\n"
"\n"
"  Minimum intron length. The default is 70 bases.\n"
"\n"
"-I INT,  --max-intron_length=INT\n"
"\n"
"  Maximum intron length. The default is 500000 bases.\n"
"\n"
"-s STRING, --strandness=STRING\n"
"\n"
"  Strand specificity of RNA library preparation. Use UN for unstranded,\n"
"  RF for first-strand, FR for second-strand (case-insensitive). The\n"
"  default is UN. This is used to set the strand in the junction-format\n"
"  BED file.\n"
"\n"
"-j FILE, --junction-bed=FILE\n"
"\n"
"  Write the junction calls and support anchors to this file. This is in\n"
"  the same format as ToHat junctions.bed and regtools junction extract\n"
"  output. It UCSC BED track, with each junction consists of two\n"
"  connected BED blocks, where each block is as long as the maximal\n"
"  overhang of any read spanning the junction. The score is the number of\n"
"  alignments spanning the junction.\n"
"\n"
"-n FILE, --intron-bed=FILE\n"
"\n"
"  Write the intron BED with the bounds of the intron. The score is the\n"
"  number of alignments spanning the junction.\n"
"\n"
"BUGS\n"
"====\n"
"\n"
"See GitHub Issues: https://github.com/diekhans/intronProspector/issues\n"
"\n"
"AUTHOR\n"
"======\n"
"\n"
"Mark Diekhans markd@ucsc.edu\n"
"\n"
"Base on code from RegTools https://github.com/griffithlab/regtools by\n"
"Avinash Ramu aramu@genome.wustl.edu.\n"
"\n"
